rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,485 @@
+   as opposed to the compressed tarballs our current build process produces, and
+3. The checksum files must be signed with a GPG key
+
+### Proposed Implementation
+
+Although the Provider Registry Specification is written in terms of a web
+service that will dynamically generate the appropriate responses, it is
+_entirely possible_ to generate a registry index as a simple directory tree and
+then statically serve those files to a client. This approach dramatically
+simplifies the process of applying updates to the registry, and is therefore
+the approach we will consider here.
+
+#### Why can't we just use Houston?
+
+The existing Houston distribution tool at [get.gravitational.com](https://get.gravitational.com/)
+is not suitable for distributing the files indexed by the registry. Houston has
+_very_ particular requirements about the naming of the files it serves (e.g.
+all files are expected to be `*.tar.gz` compressed tarballs - or at least
+_named_ as such) and, as it stands, will not serve the files we need it to.
+
+We _could_ modify Houston to serve the required files, but as the service is
+already slated for retirement, that does not seem to be a good use of time
+& money.
+
+### Overall architecture
+
+In short:
+
+* Separate _Staging_ and _Production_ registries.
+  * _Staging_ will serve pre-release provider packages, and will be updated on
+    every tag build,
+  * _Production_ will serve released provider versions, and will be updated
+    via promoting a specific tag build to production via Drone
+* Each registry will be stored by an AWS S3 bucket and served by AWS CloudFront
+* The storage bucket will be configured as per the Gold-class storage requirements
+described in the [Cloud Artifact Storage Standards](https://github.com/gravitational/cloud/blob/master/rfd/0017-artifact-storage-standards.md)
+RFD.
+
+#### Public naming
+
+##### Staging
+
+As the the staging registry is only intended for internal testing, we _could_
+simply use the CloudFront-assigned public domain name as the publicly-visible
+domain of the staging registry (e.g. `d1cqyxzpzj86o0.cloudfront.net`).
+
+We can also add an alternate domain name (e.g. `terraform.staging.teleport.dev`)
+by
+
+1. adding an alias record to our DNS config that points
+   `terraform.staging.teleport.dev` back to the CloudFront CCN, and
+2. configuring CloudFront to use a custom TLS certificate when serving
+  requests for `terraform.staging.teleport.dev`.
+
+##### Production
+
+In selecting the name for the production registry, there is a tension between
+having (arguably) the most memorable and obvious name for the registry (but
+having it's configuration split of several hosts), versus and having a registry
+with all of its configuration and data in one place.
+
+How we name the production registry ultimately comes down to deciding which
+side of this tension we care more about.
+
+If we want people to reference our provider as `goteleport.com/gravitational/teleport`,
```

Comment:
On that note though, how would we disambiguate a docker image and a terraform provider that would presumably share the same naming scheme.

Or potentially helm charts. I personally think `terraform.releases.teleport.dev/gravitational/teleport` is a very lengthy name, but `goteleport.com/gravitational/teleport` could potentially have multiple artifacts vying for that spot. 

Thoughts on `terraform.goteleport.com/gravitational/teleport`?

Reply:
I'd be happy with that, certainly. I've only suggested the longer `terraform.releases.teleport.dev` as it matches the existing naming scheme. I'll add it to the list.

Reply:
Updated